Extension content scripts are not exempted from Trusted Types page CSP

nt1m opened the pull request:
#### 421a4d041cbc7743843f5c824f205c0ca017cb98
<pre>
Extension content scripts are not exempted from Trusted Types page CSP
<a href="https://bugs.webkit.org/show_bug.cgi?id=290037">https://bugs.webkit.org/show_bug.cgi?id=290037</a>
<a href="https://rdar.apple.com/147411842">rdar://147411842</a>

Reviewed by Timothy Hatcher and Ryosuke Niwa.

Extension content scripts run into their own world, and webpages that use a Trusted Types CSP should not affect content scripts written by extensions.

Change `Document::requiresTrustedTypes` to only enforce the checks when the script executing is in the main world (the document&apos;s), to exclude
extension content scripts.

Credits to Timothy Hatcher for the test.

* Source/WebCore/dom/Document.h:
(WebCore::Document::requiresTrustedTypes const):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIScripting.mm:
(TestWebKitAPI::TEST(WKWebExtensionAPIScripting, InjectScriptWithTrustedTypesCSP)):

Canonical link: <a href="https://commits.webkit.org/292473@main">https://commits.webkit.org/292473@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9dc06989444af9ccee8d3e945c095a051bfa5824

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/95881 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/15495 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/5436 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/100940 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/46396 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/97926 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/15790 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/23927 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/73111 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/30341 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/98884 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/11826 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/86608 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/53439 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/11557 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/4352 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/45725 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/81728 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/4458 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/102969 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/22948 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/16734 "Found 2 new test failures: html5lib/generated/run-entities01-write.html imported/w3c/web-platform-tests/html/infrastructure/safe-passing-of-structured-data/shared-array-buffers/no-coop-coep.https.any.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/82153 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/23199 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/82628 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/81512 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20507 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/26100 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/3552 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/16285 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/22911 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/28066 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/22570 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/26049 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/24311 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/dom/Document.h:
```diff
@@ -853,7 +853,7 @@ class Document
     void setTrustedTypesEnforcement(JSC::TrustedTypesEnforcement) final;
 
-    bool requiresTrustedTypes() const { return m_requiresTrustedTypes; }
+    bool requiresTrustedTypes() const { return m_requiresTrustedTypes && !shouldBypassMainWorldContentSecurityPolicy(); }
```
Hm... this function is called in pretty hot functions like `Element::setAttribute` and alike.
We should probably schedule perf A/B for speedometer.

lukewarlow replied:
It might be better to add this check when we set the value rather than access it?

rniwa replied:
We can't do that because shouldBypassMainWorldContentSecurityPolicy looks up the call stack to determine which world is used to run the script.

lukewarlow commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIScripting.mm:
```diff
@@ -1655,6 +1655,43 @@
+
+    auto *contentScript = Util::constructScript(@[
+        @"try {",
+        @"  document.body.innerHTML = '<p>Injected</p>'",
```
Is it worth adding a separate test for eval within an extension? (Is that's allowed)

xeenon replied:
`eval()` is blocked for other reasons in extension content scripts — it isn't allowed in manifest v3 anymore.

nt1m committed:
Extension content scripts are not exempted from Trusted Types page CSP
https://bugs.webkit.org/show_bug.cgi?id=290037
rdar://147411842

Reviewed by Timothy Hatcher and Ryosuke Niwa.

Extension content scripts run into their own world, and webpages that use a Trusted Types CSP should not affect content scripts written by extensions.

Change `Document::requiresTrustedTypes` to only enforce the checks when the script executing is in the main world (the document's), to exclude
extension content scripts.

Credits to Timothy Hatcher for the test.

* Source/WebCore/dom/Document.h:
(WebCore::Document::requiresTrustedTypes const):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIScripting.mm:
(TestWebKitAPI::TEST(WKWebExtensionAPIScripting, InjectScriptWithTrustedTypesCSP)):

Canonical link: https://commits.webkit.org/292473@main
